docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,413 @@
+
+## Alternative Approaches
+
+While the adoption of ABCI++ is inevitable given the already made decision by upstream to implement it, here are some alternatives to the features that we need that do not use ABCI++:
+
+- [Alternatives for Message Inclusion.](https://github.com/celestiaorg/celestia-app/blob/92341dd68ee6e555ec6c0bb780afa3a1c8243a93/adrs/adr008:adopt-ABC%2B%2B-early.md#alternative-approaches)
+- [Alternatives for Picking a square size.](https://github.com/celestiaorg/celestia-core/issues/454)
+
+## Decision
+
+Proposed and initial implementation is complete.
+
+## Detailed Design
+
+### [#631](https://github.com/celestiaorg/celestia-core/pull/631) Simplified version of ABCI++ (`celestia-core`)
+
+Here we are adding only the two new methods that are necessary for the features that we need.  
+```go
+// Application is an interface that enables any finite, deterministic state machine
+// to be driven by a blockchain-based replication engine via the ABCI.
+type Application interface {
+   ...
+   PrepareProposal(RequestPrepareProposal) ResponsePrepareProposal
+   ProcessProposal(RequestProcessProposal) ResponseProcessProposal
+   ...
+}
+```
+
+It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data. This is because Celestia separates some block data that cannot modify state (messages), and the application has to have access to both normal transaction data and messages to perform the necessary processing and checks.
+```protobuf
+message RequestPrepareProposal {
+ // block_data is an array of transactions that will be included in a block,
+ // sent to the app for possible modifications.
+ // applications can not exceed the size of the data passed to it.
+ tendermint.types.Data block_data = 1;
+ // If an application decides to populate block_data with extra information, they can not exceed this value.
+ int64 block_data_size = 2;
+}
+
+message RequestProcessProposal {
+ tendermint.types.Header header     = 1 [(gogoproto.nullable) = false];
+ tendermint.types.Data   block_data = 2;
+}
+
+...
+
+// Data contains the set of transactions, evidence, intermediate state roots,
+// and messages to be included in the block
+message Data {
+ // Txs that will be applied by state @ block.Height+1.
+ // NOTE: not all txs here are valid.  We're just agreeing on the order first.
+ // This means that block.AppHash does not include these txs.
+ repeated bytes txs = 1;
+
+ IntermediateStateRoots intermediate_state_roots = 2 [(gogoproto.nullable) = false];
```

Comment:
I'd already skip it here tbh.
Add a comment that field number to is reserved for ISRs instead.

Reply:
good idea :+1: https://github.com/celestiaorg/celestia-app/pull/231/commits/9dace036b5c51076f73e8c38699b5360b11ee3a1